server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/matchers/Xor.java

```diff
@@ -42,29 +43,27 @@
 public class Xor extends GenericCompositeMatcher {
 
     @Override
-    public Collection<MailAddress> match(Mail mail) throws MessagingException {
-        Collection<MailAddress> finalResult = null;
-        boolean first = true;
-        for (Matcher matcher: getMatchers()) {
-            Collection<MailAddress> matchedAddresses = Optional.ofNullable(matcher.match(mail)).orElse(new ArrayList<>());
+    public Collection<MailAddress> match(Mail mail) {
+        Set<MailAddress> recipients = getMatchers().stream()
+            .map(Throwing.<Matcher, Set<MailAddress>>function(matcher -> applyMatcher(mail, matcher)).sneakyThrow())
+            .reduce(ImmutableSet.of(), this::performXor);
 
-            if (first) {
-                finalResult = matchedAddresses;
-                first = false;
-            } else {
-                finalResult = performXor(finalResult, matchedAddresses);
-            }
+        if (recipients.isEmpty()) {
+            return null;
```

Comment:
why not empty list? though

Reply:
because I don't intend to change the existing behavior (and it should be a rule when doing some refactoring)